content/blog/callpackage-a-tool-for-the-lazy.md

```diff
@@ -0,0 +1,219 @@
+---
+title: callPackage, a tool for the lazy
+date: 2022-08-25
+extra:
+  author: Norbert Melzer
+---
+
+In [`nixpkgs`](https://github.com/nixos/nixpkgs), there is a massive use of the
+`callPackage` function, which provides us with a lot of benefits.
+
+## Basic examples
+
+Though before even discussing the benefits, lets see how it actually gets used.
+
+Given are these 2 files `hello.nix` and `default.nix`:
+
+```nix
+# Usually I would prefer the usage of nix flakes, though that would
+# introduce a lot of boilerplate in the examples for no reason,
+# therefore I will use diamond paths and `import` for simplicity.
+```
+
+```nix
+# default.nix
+let pkgs = import <nixpkgs> {}; in
+pkgs.callPackage ./hello.nix {}
+```
+
+```nix
+# hello.nix
+{writeShellScriptBin}:
+writeShellScriptBin "hello" ''
+  echo "hello, world!"
+''
+```
+
+Building using `nix-build` will produce `./result/bin/hello`, and running it will
+nicely greet you.
+
+As you can see, `writeShellScriptBin` gets passed in by `callPackage` automatically.
+
+For this simple setup, having to create an extra file, seems to be a lot of
+boilerplate, though if you continue reading, you will see, it is worth it!
```

Comment:
```suggestion
It may appear cumbersome to create an extra file for the package in such a simple setup.
In fact, this is exactly how `nixpkgs` is organized: every package is a file that declares a function.
This function takes as arguments the package's dependencies.

If you continue reading, you will see the benefits of this pattern!
```